### DISCUSSION TITLE
Unable to build for Android using termux

### DISCUSSION THREAD

**Discussion by agret** [ggml-org/llama.cpp#8737] [Q&A]
When trying to build for Android:
![Screenshot_2024-07-28-21-00-08-36_84d3000e3f4017145260f7618db1d683](https://github.com/user-attachments/assets/52c0e675-5e0c-42cd-97c9-b754a3ee5259)
I was following this guide: https://github.com/JackZeng0208/llama.cpp-android-tutorial

I found execinfo.h is not included in the Android NDK so something must've been changed with the project in the 10 months since that guide was written to stop it building for Android.

Then I follow for your guide:
https://github.com/ggerganov/llama.cpp/blob/master/docs/android.md
![Screenshot_2024-07-28-22-48-18-55_84d3000e3f4017145260f7618db1d683](https://github.com/user-attachments/assets/b7eec80e-5436-4d06-9bc5-9a6d3db32484)
As I thought, trying to run a cmake from an empty directory didn't work. Seems there are some steps missing.

Where can I get the make files for an Android build?

**Comment by myan-o** [comment]
mkdir build
cd build
cmake ..
make -j$(nproc)
bin/llama-server -m model_path

**Comment by mugenyume** [comment]
> When trying to build for Android: ![Screenshot_2024-07-28-21-00-08-36_84d3000e3f4017145260f7618db1d683](https://private-user-images.githubusercontent.com/568523/352812996-52c0e675-5e0c-42cd-97c9-b754a3ee5259.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI0NjIzNDEsIm5iZiI6MTcyMjQ2MjA0MSwicGF0aCI6Ii81Njg1MjMvMzUyODEyOTk2LTUyYzBlNjc1LTVlMGMtNDJjZC05N2M5LWI3NTRhM2VlNTI1OS5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzMxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDczMVQyMTQwNDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02NmM5ODYxNWM4NjhiNzUxNTkwZjdmNTY1YWNlM2YwNjFlY2M5ZGM5YmFhNjJmNGMzZGMyNDkwOWExZmQ3NjE3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.7jkh5Bk2gTI2nQ_jUXjlbfCHmh_-DGhSp8J3HB9F5Kg) I was following this guide: https://github.com/JackZeng0208/llama.cpp-android-tutorial
> 
> I found execinfo.h is not included in the Android NDK so something must've been changed with the project in the 10 months since that guide was written to stop it building for Android.
> 
> Where can I get the make files for an Android build?

To fix this issue, follow these steps to download and set up the Android NDK for Termux:

1. Go to the [Termux Android NDK GitHub page](https://github.com/lzhiyong/termux-ndk).

2. Download the NDK:
   - Either download it directly from your browser and then import it into Termux.
   - Or use the command `wget <link>` in Termux to download it.

3. Unzip the downloaded file by executing the following command:
   ```sh
   unzip <file.zip>
   ```

4. Set the NDK directory by executing the following command:
   ```sh
   export NDK=<your_ndk_directory>
   ```

I had the same problem when I tried to build Ollama with Go, and this solution resolved my issue.

Try building it again, it should work rn. 
Hope this solves your issue also! :)

**Reply by mugenyume** [reply]
By the way, to ensure everything works smoothly, I also installed the following dependencies:

```sh
pkg install build-essential cmake git automake patchelf libexpat openssl libandroid-execinfo ninja protobuf libsodium
```

So, if the steps I mentioned above don't work, try installing these dependencies and then try again.
